src/systems/audio-debug-system.js

```diff
@@ -0,0 +1,323 @@
+      background = mix(background, outerLayer2, outerLayer2.a * (clipped[i] ? 0.0 : 1.0));
+
+      // Draw base
+      vec4 baseLayer = circle(center, d, 1.0, 1.0, 0.1, vec3(0.5, 0.5, 0.5), 0.0);
+      background = mix(background, baseLayer, baseLayer.a);
+
+      // Draw gain
+      float g = clamp(gain[i], 0.0, 1.0);
+      vec4 gainLayer = circle(center, d, g, 1.0, 0.5, colorGain, startOffset);
+      background = mix(background, gainLayer, gainLayer.a);
+      if (gain[i] > 1.0) {
+        vec4 overGainLayer = circle(center, d, gain[i] - g, 1.0, 0.5, vec3(1.0, 0.0, 0.0), startOffset);
+        background = mix(background, overGainLayer, overGainLayer.a);
+      }
+    }
+    
+    // Blend
+    gl_FragColor = vec4(background.rgb, background.a * vNormal.y);
+  }
+`;
+
+AFRAME.registerSystem("audio-debug", {
+  schema: {
+    enabled: { default: false }
+  },
+
+  init() {
+    window.APP.store.addEventListener("statechanged", this.updateState.bind(this));
+
+    this.sources = [];
+
+    this.material = new THREE.ShaderMaterial({
+      uniforms: {
+        time: { value: 0.0 },
+        colorInner: { value: new THREE.Color("#7AFF59") },
+        colorOuter: { value: new THREE.Color("#FF6340") },
+        colorGain: { value: new THREE.Color("#70DBFF") },
+        count: { value: 0 },
+        maxDistance: { value: [] },
+        refDistance: { value: [] },
+        rolloffFactor: { value: [] },
+        distanceModel: { value: [] },
+        sourcePosition: { value: [] },
+        sourceOrientation: { value: [] },
+        coneInnerAngle: { value: [] },
+        coneOuterAngle: { value: [] },
+        gain: { value: [] },
+        clipped: { value: [] }
+      },
+      vertexShader: CONE_VERTEX,
+      fragmentShader: CONE_FRAG
+    });
+    this.material.side = THREE.FrontSide;
+    this.material.transparent = true;
+    this.material.uniforms.count.value = 0;
+
+    this.sourcePositions = new Array(MAX_DEBUG_SOURCES);
+    this.sourceOrientations = new Array(MAX_DEBUG_SOURCES);
+    this.distanceModels = new Array(MAX_DEBUG_SOURCES);
+    this.maxDistances = new Array(MAX_DEBUG_SOURCES);
+    this.refDistances = new Array(MAX_DEBUG_SOURCES);
+    this.rolloffFactors = new Array(MAX_DEBUG_SOURCES);
+    this.coneInnerAngles = new Array(MAX_DEBUG_SOURCES);
+    this.coneOuterAngles = new Array(MAX_DEBUG_SOURCES);
+    this.gains = new Array(MAX_DEBUG_SOURCES);
+    this.clipped = new Array(MAX_DEBUG_SOURCES);
+  },
+
+  remove() {
+    window.APP.store.removeEventListener("statechanged", this.updateState);
+  },
+
+  registerSource(source) {
+    this.sources.push(source);
+  },
+
+  unregisterSource(source) {
+    const index = this.sources.indexOf(source);
+
+    if (index !== -1) {
+      this.sources.splice(index, 1);
+    }
+  },
+
+  tick(time) {
+    if (!this.data.enabled) {
+      return;
+    }
+
+    this.sourcePositions.fill(new THREE.Vector3());
```

Comment:
Same as mentioned elsewhere, avoid creating threejs objects every frame. For these in particular you are actually copying the reference anyway so I don't think you even need to update these every frame, just when a new source is added or removed.

Reply:
Yep, I've moved all that to the initalization.